### PR TITLE
Construct query job latches on-demand

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister

### CODE REVIEW

**Comment by Zoxc** [comment]
The idea here is to avoid creating a `Lrc<QueryJob>` each time we want to execute a query and instead create a `QueryLatch = Lrc<Mutex<QueryLatchInfo<'tcx>>>` when we actually want to wait for a query.

A `QueryJob` is now stored inline in the `active` map and contains a `Option<QueryLatch>` which is created on demand. This way we avoid the allocation and the atomic operations when there's no waiters on the query.

We can no longer refer to queries using `Lrc<QueryJob>` which is a small type. Using `Query<'tcx>` is too expensive as that is quite big. Instead `QueryToken` is introduced which is a pointer to a stack value which uniquely identifies an active query execution. A `QueryToken` is stored for each running query in the `active` map. This allows you to look up information for a token by looking at all the `active` maps. While this is expensive, it only happens on cycle errors and ICEs.

**Comment by michaelwoerister** [comment]
The performance numbers look good. I'll try to make time for reviewing this some time this or next week.

File: src/librustc/ty/query/job.rs [hunk 1]
```diff
@@ -30,61 +31,119 @@ pub struct QueryInfo<'tcx> {
...
 
-/// Representss an object representing an active query job.
-pub struct QueryJob<'tcx> {
+type QueryMap<'tcx> = FxHashMap<QueryToken, QueryJobInfo<'tcx>>;
+
+/// A value uniquely identifiying an active query job.
+/// This value is created from a stack pointer in `get_query` and `force_query`
+/// which is alive while the query executes.
+#[derive(Copy, Clone, Eq, PartialEq, Hash)]
+pub struct QueryToken(NonZeroUsize);
```
Comment (michaelwoerister) [inline src/librustc/ty/query/job.rs]: Using a pointer into the stack as a unique identifier is clever but I'm a bit worried that we can't make the compiler guarantee that the value won't be reused. The core logic within the query engine is very complicated and somebody might do some kind of optimization that silently breaks the invariant of the `key` stack value being alive throughout the query jobs lifetime.

If I understand things correctly, the requirements for `QueryToken` are:
 - it needs to uniquely identify a QueryJob
 - it needs to be cheap to construct.

At construction time we already have a lock on the query cache. Could we put a `u64` counter into that?

Also, if the main purpose of `QueryToken` is to identify a `QueryJob` then I'd prefer if it was called `QueryJobId`.

**Review by michaelwoerister — COMMENTED** [review]
I'm generally in favor of going forward with this. It's a good idea to move some of the more expensive operations out of the fast-path. I have one main concern that I listed below.

File: src/librustc/ty/query/job.rs (hunk 1 above)
Comment (Zoxc) [reply src/librustc/ty/query/job.rs]: We can put a counter into the query cache shard. We'll need to split `QueryToken` into `QueryJobId` (which globally identifies a query) and `QueryShardJobId` (which just identifies a query within the shard), or we could just use `u32` for the latter.

Comment (michaelwoerister) [reply src/librustc/ty/query/job.rs]: I was thinking just having a u64 counter per shard and initializing each counter in a way that the upper N bits globally identify the shard. E.g. allocate a 20 bit "(query-kind, shard) id" from a static atomic counter when the shard is created and then initialize the shard-local counters with `shard_id << 44`. But I don't actually care about the implementation details as long as `QueryJobId` does what its name suggests.

Comment (Zoxc) [reply src/librustc/ty/query/job.rs]: I used `DepKind`, `u16` (for the shard) and `u32` for a job in the shard. We're limited to `u32::MAX - 1` due to `DepNodeIndex` anyway.

File: src/librustc/ty/query/plumbing.rs [hunk 2]
```diff
@@ -139,39 +149,54 @@ impl<'a, 'tcx, Q: QueryDescription<'tcx>> JobOwner<'a, 'tcx, Q> {
                                 query_blocked_prof_timer = Some(tcx.prof.query_blocked());
                             }
 
-                            job.clone()
+                            // Create the id of the job we're waiting for
+                            let id = QueryJobId {
```
Comment (michaelwoerister) [inline src/librustc/ty/query/plumbing.rs]: Would you mind creating a helper method for this? This function is already rather big.

File: src/librustc/ty/query/plumbing.rs [hunk 3]
```diff
@@ -139,39 +149,54 @@ impl<'a, 'tcx, Q: QueryDescription<'tcx>> JobOwner<'a, 'tcx, Q> {
...
+                        // Generate an id unique within this shard.
+                        let id = jobs.checked_add(1).unwrap();
+                        *jobs = id;
+                        let id = QueryShardJobId(NonZeroU32::new(id).unwrap());
+
+                        let global_id = QueryJobId {
+                            job: id,
+                            shard: u16::try_from(shard).unwrap(),
+                            kind: Q::dep_kind(),
+                        };
```
Comment (michaelwoerister) [inline src/librustc/ty/query/plumbing.rs]: Is there are reason for not creating the `global_id` outside the closure. That seems simpler than capturing the `jobs` reference.

File: src/librustc/ty/query/plumbing.rs [hunk 4]
```diff
@@ -81,12 +88,7 @@ impl<'a, 'tcx, Q: QueryDescription<'tcx>> JobOwner<'a, 'tcx, Q> {
     /// This function is inlined because that results in a noticeable speed-up
     /// for some compile-time benchmarks.
     #[inline(always)]
-    pub(super) fn try_get(
-        tcx: TyCtxt<'tcx>,
-        span: Span,
-        key: &Q::Key,
-        token: QueryToken,
-    ) -> TryGetJob<'a, 'tcx, Q> {
+    pub(super) fn try_get(tcx: TyCtxt<'tcx>, span: Span, key: &Q::Key) -> TryGetJob<'a, 'tcx, Q> {
```
Comment (michaelwoerister) [inline src/librustc/ty/query/plumbing.rs]: I like how `token` disappears from the function signature again.

File: src/librustc/ty/query/job.rs (hunk 1 above)
Comment (michaelwoerister) [reply src/librustc/ty/query/job.rs]: Yes, I like how that turned out.

File: src/librustc/ty/query/plumbing.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc/ty/query/plumbing.rs]: Done.

File: src/librustc/ty/query/plumbing.rs (hunk 3 above)
Comment (Zoxc) [reply src/librustc/ty/query/plumbing.rs]: No, this is was my default way of dealing with this particular borrow error, I've refactored it.

**Comment by michaelwoerister** [comment]
Thanks, @Zoxc!

@bors r+ rollup=never

(This is doing some complicated changes, let's keep it out of rollups in case it breaks anything)

**Pull request merged**